### PR TITLE
Handle disposal of a HeroState while a hero is animating

### LINKED ISSUE #5178: HeroState._setChild exception in scheduler callback after popping route
1) Run Gallery and go to Pesto
2) Scroll the recipe list
3) Select Home in the menu
4) Tap the back button

This produces the exception below.

══╡ EXCEPTION CAUGHT BY SCHEDULER LIBRARY ╞═════════════════════════════════════════════════════════
The following assertion was thrown during a scheduler callback:
'package:flutter/src/widgets/heroes.dart': Failed assertion: line 193: 'mounted' is not true.
When the exception was thrown, this was the stack:
#2      HeroState._setChild (package:flutter/src/widgets/heroes.dart:193)
#3      HeroParty._handleUpdate (package:flutter/src/widgets/heroes.dart:393)
#4      Animation&AnimationLazyListenerMixin&AnimationLocalStatusListenersMixin.notifyStatusListeners (package:flutter/src/animation/listener_helpers.dart:118)
#5      ReverseAnimation._statusChangeHandler (package:flutter/src/animation/animations.dart:276)
#6      Animation&AnimationLazyListenerMixin&AnimationLocalListenersMixin&AnimationLocalStatusListenersMixin.notifyStatusListeners (package:flutter/src/animation/listener_helpers.dart:118)
#7      Animation&AnimationEagerListenerMixin&AnimationLocalListenersMixin&AnimationLocalStatusListenersMixin.notifyStatusListeners (package:flutter/src/animation/listener_helpers.dart:118)
#8      AnimationController._checkStatusChanged (package:flutter/src/animation/animation_controller.dart:240)
#9      AnimationController._tick (package:flutter/src/animation/animation_controller.dart:251)
#10     Ticker._tick (package:flutter/src/scheduler/ticker.dart:77)
#11     BindingBase&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:488)
#12     BindingBase&SchedulerBinding._invokeTransientFrameCallbacks.<anonymous closure> (package:flutter/src/scheduler/binding.dart:471)
#13     _HashVMBase&MapMixin&&_LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:345)
#14     BindingBase&SchedulerBinding._invokeTransientFrameCallbacks (package:flutter/src/scheduler/binding.dart:469)
#15     BindingBase&SchedulerBinding.handleBeginFrame (package:flutter/src/scheduler/binding.dart:445)
#16     _beginFrame (file:///ssd2/sky/engine/src/out/android_debug_unopt/gen/sky/bindings/dart_ui/hooks.dart:48)
(elided 2 frames from class _AssertionError)
This exception was thrown in the context of a scheduler callback. When the scheduler callback was
_registered_ (as opposed to when the exception was thrown), this was the stack:
#0      _FrameCallbackEntry._FrameCallbackEntry.<anonymous closure> (package:flutter/src/scheduler/binding.dart:76)
#2      _FrameCallbackEntry._FrameCallbackEntry (package:flutter/src/scheduler/binding.dart:57)
#3      BindingBase&SchedulerBinding.addFrameCallback (package:flutter/src/scheduler/binding.dart:250)
#4      BindingBase&SchedulerBinding.scheduleFrameCallback (package:flutter/src/scheduler/binding.dart:219)
#5      Ticker._scheduleTick (package:flutter/src/scheduler/ticker.dart:87)
#6      Ticker.start (package:flutter/src/scheduler/ticker.dart:42)
#7      AnimationController._startSimulation (package:flutter/src/animation/animation_controller.dart:217)
#8      AnimationController.animateTo (package:flutter/src/animation/animation_controller.dart:182)
#9      AnimationController.reverse (package:flutter/src/animation/animation_controller.dart:162)
#10     TransitionRoute.didPop (package:flutter/src/widgets/routes.dart:185)
#11     TransitionRoute&&LocalHistoryRoute.didPop (package:flutter/src/widgets/routes.dart:330)
#12     NavigatorState.pop (package:flutter/src/widgets/navigator.dart:442)
#13     _WidgetsAppState.didPopRoute (package:flutter/src/widgets/app.dart:134)
#14     BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding&WidgetsBinding.handlePopRoute (package:flutter/src/widgets/binding.dart:163)
#15     _popRoute (file:///ssd2/sky/engine/src/out/android_debug_unopt/gen/sky/bindings/dart_ui/hooks.dart:38)
(elided one frame from class _AssertionError)

### PR DESCRIPTION
**Pull request by jason-simmons**
Fixes https://github.com/flutter/flutter/issues/5178

### CODE REVIEW

**Comment by jason-simmons** [comment]
@abarth @Hixie

**Comment by Hixie** [comment]
@abarth, since i helped write this it'd be best if you reviewed it

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 1]
```diff
@@ -279,6 +306,16 @@ class _HeroQuestState implements HeroHandle {
       )
     );
   }
+
+  void dispose() {
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/heroes.dart]: `@mustCallSuper` just in case anyone subclasses.

**Comment by abarth** [comment]
Looks really complicated.
LGTM

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 1 above)
Comment (jason-simmons) [inline packages/flutter/lib/src/widgets/heroes.dart]: done

**Pull request merged**